File: populator.py
import logging
import random
from collections import defaultdict
from mysql.connector import Error

class DatabasePopulator:
    """
    Populates database tables with fake data
    """
    def __init__(self, db_connector, schema_analyzer, data_generator, num_records=10, max_retries=5):
        """
        Initialize with database connector, schema analyzer, and data generator

        Args:
            db_connector: DatabaseConnector instance
            schema_analyzer: SchemaAnalyzer instance
            data_generator: DataGenerator instance
            num_records (int): Number of records to generate per table
            max_retries (int): Maximum number of retries for handling circular dependencies
        """
        self.db = db_connector
        self.schema = schema_analyzer
        self.data_gen = data_generator
        self.num_records = num_records
        self.max_retries = max_retries
        self.inserted_data = defaultdict(list)
        self.failed_tables = set()

    def populate_database(self):
        """
        Populate all tables in the database

        Returns:
            bool: Success status
        """
        # Analyze schema
        if not self.schema.analyze_schema():
            logging.error("Failed to analyze database schema")
            return False

        # Get table insertion order
        ordered_tables, circular_tables = self.schema.get_table_insertion_order()

        if not ordered_tables:
            logging.error("No tables to populate")
            return False

        logging.info(f"Tables will be populated in the following order: {', '.join(ordered_tables)}")
        logging.info(f"Tables involved in circular dependencies: {', '.join(circular_tables)}")

        # Special handling for SyncSource and SyncSource_LogAlert
        # Make sure SyncSource is populated first
        if 'SyncSource' in ordered_tables and 'SyncSource_LogAlert' in ordered_tables:
            logging.info("Special handling for SyncSource and SyncSource_LogAlert tables")

            # Remove these tables from the ordered list
            ordered_tables = [t for t in ordered_tables if t not in ['SyncSource', 'SyncSource_LogAlert']]

            # Add SyncSource to the beginning of the list
            ordered_tables = ['SyncSource'] + ordered_tables

            # Add SyncSource_LogAlert to the end of the list
            ordered_tables.append('SyncSource_LogAlert')

        # First pass: populate tables without circular dependencies
        non_circular_tables = [table for table in ordered_tables if table not in circular_tables]
        for table in non_circular_tables:
            self._populate_table(table)

        # Second pass: populate tables with circular dependencies
        if circular_tables:
            logging.info("Starting second pass to populate tables with circular dependencies")

            # Try to populate each table in the circular dependency
            for table in [t for t in ordered_tables if t in circular_tables]:
                self._populate_table(table, handle_circular=True)

            # Check if any tables failed to populate
            if self.failed_tables:
                logging.warning(f"Some tables could not be fully populated: {', '.join(self.failed_tables)}")

                # Try random table order for failed tables
                self._handle_failed_tables()

        # Final check
        success = len(self.failed_tables) == 0
        if success:
            logging.info("All tables successfully populated")
        else:
            logging.warning(f"The following tables could not be fully populated: {', '.join(self.failed_tables)}")

        return success

    def _populate_table(self, table, handle_circular=False):
        """
        Populate a single table with fake data

        Args:
            table (str): Table name
            handle_circular (bool): Whether to handle circular dependencies

        Returns:
            bool: Success status
        """
        if table not in self.schema.table_columns:
            logging.error(f"Table {table} not found in schema")
            self.failed_tables.add(table)
            return False

        logging.info(f"Populating table: {table}")

        # Get column information
        columns = self.schema.table_columns[table]
        column_names = [col['column_name'] for col in columns]

        # Special handling for SyncSource table which has a column named 'Lag' (MySQL reserved keyword)
        if table == 'SyncSource':
            logging.info("Special handling for SyncSource table with reserved keyword columns")
            # Escape column names with backticks to handle reserved keywords
            escaped_column_names = [f"`{col_name}`" for col_name in column_names]

            # Prepare placeholders for SQL query
            placeholders = ', '.join(['%s'] * len(column_names))

            # Build INSERT query with escaped column names
            insert_query = f"INSERT INTO `{table}` ({', '.join(escaped_column_names)}) VALUES ({placeholders})"

            logging.debug(f"Using escaped column names for SyncSource: {insert_query}")
        else:
            # Prepare placeholders for SQL query
            placeholders = ', '.join(['%s'] * len(column_names))

            # Build INSERT query
            insert_query = f"INSERT INTO {table} ({', '.join(column_names)}) VALUES ({placeholders})"

        # Generate and insert data
        successful_inserts = 0
        for i in range(self.num_records):
            try:
                # Generate row data
                row_data = self._generate_row_data(table, columns, handle_circular)

                # Execute insert
                if self.db.execute_query(insert_query, tuple(row_data.values()), commit=True):
                    # Get the inserted ID if there's an auto-increment column
                    last_id = None
                    for col in columns:
                        if col['extra'] == 'auto_increment':
                            last_id_query = "SELECT LAST_INSERT_ID() as id"
                            result = self.db.execute_query(last_id_query)
                            if result and result[0]['id']:
                                last_id = result[0]['id']
                                row_data[col['column_name']] = last_id

                    # Store inserted data for foreign key references
                    self.inserted_data[table].append(row_data)
                    successful_inserts += 1

            except Error as e:
                logging.error(f"Error inserting into {table}: {e}")
                # Continue with next record

        # Check if all records were inserted successfully
        if successful_inserts < self.num_records:
            logging.warning(f"Only {successful_inserts}/{self.num_records} records inserted into {table}")
            if successful_inserts == 0:
                self.failed_tables.add(table)
                return False

        return True

    def _generate_row_data(self, table, columns, handle_circular=False):
        """
        Generate data for a single row

        Args:
            table (str): Table name
            columns (list): List of column information dictionaries
            handle_circular (bool): Whether to handle circular dependencies

        Returns:
            dict: Generated row data
        """
        row_data = {}

        # Special handling for SyncSource_LogAlert table which has a circular dependency with SyncSource
        if table == 'SyncSource_LogAlert':
            # First, check if we have any SyncSource records in the database
            query = "SELECT ID FROM SyncSource LIMIT 1"
            result = self.db.execute_query(query)

            if result and len(result) > 0:
                # We have SyncSource records, so we can use them
                logging.info("Found existing SyncSource records to use for SyncSource_LogAlert")

                # Get all SyncSource IDs
                query = "SELECT ID FROM SyncSource"
                result = self.db.execute_query(query)

                if result and len(result) > 0:
                    # Store these IDs for use in the foreign key
                    sync_source_ids = [row['ID'] for row in result]

                    # Remember this for later when we process the SyncSource_ID column
                    self._sync_source_ids = sync_source_ids

        # Process columns
        for col in columns:
            column_name = col['column_name']

            # Special handling for SyncSource_LogAlert.SyncSource_ID
            if table == 'SyncSource_LogAlert' and column_name == 'SyncSource_ID' and hasattr(self, '_sync_source_ids') and self._sync_source_ids:
                # Use an existing SyncSource ID
                row_data[column_name] = random.choice(self._sync_source_ids)
                logging.info(f"Using existing SyncSource ID {row_data[column_name]} for SyncSource_LogAlert.SyncSource_ID")
                continue

            # Check if this is a foreign key
            is_foreign_key = False
            referenced_table = None
            referenced_column = None

            if table in self.schema.foreign_keys:
                for fk in self.schema.foreign_keys[table]:
                    if fk['column'] == column_name:
                        is_foreign_key = True
                        referenced_table = fk['referenced_table']
                        referenced_column = fk['referenced_column']
                        break

            # Handle foreign key
            if is_foreign_key:
                # Check if this foreign key is nullable
                is_nullable = col['is_nullable'].lower() == 'yes'

                # Find the corresponding foreign key entry to get additional info
                fk_info = None
                if table in self.schema.foreign_keys:
                    for fk in self.schema.foreign_keys[table]:
                        if fk['column'] == column_name and fk['referenced_table'] == referenced_table:
                            fk_info = fk
                            break

                # If we found the FK info, use its nullable status (which might be more accurate)
                if fk_info and 'is_nullable' in fk_info:
                    is_nullable = fk_info['is_nullable']

                # Check if referenced table has data
                if referenced_table in self.inserted_data and self.inserted_data[referenced_table]:
                    # Use an existing value from the referenced table
                    referenced_row = random.choice(self.inserted_data[referenced_table])
                    row_data[column_name] = referenced_row[referenced_column]
                elif is_nullable:
                    # This is a nullable foreign key, so we can set it to NULL
                    logging.info(f"Setting nullable foreign key {table}.{column_name} -> {referenced_table}.{referenced_column} to NULL")
                    row_data[column_name] = None
                elif handle_circular:
                    # This is a circular dependency with NOT NULL constraint
                    logging.warning(f"Circular dependency detected for {table}.{column_name} -> {referenced_table}.{referenced_column} (NOT NULL)")

                    # Try to find any value in the referenced table
                    query = f"SELECT {referenced_column} FROM {referenced_table} LIMIT 1"
                    result = self.db.execute_query(query)

                    if result and result[0][referenced_column] is not None:
                        # Use existing value from database
                        row_data[column_name] = result[0][referenced_column]
                        logging.info(f"Using existing value from database for {table}.{column_name} -> {referenced_table}.{referenced_column}")
                    else:
                        # No existing value, this is a hard circular dependency
                        # Generate a placeholder value that will be fixed later
                        placeholder = self._generate_placeholder_value(col)
                        row_data[column_name] = placeholder
                        logging.info(f"Generated placeholder value {placeholder} for {table}.{column_name} -> {referenced_table}.{referenced_column}")
                else:
                    # Referenced table not yet populated and not handling circular dependencies
                    # This should not happen if tables are properly ordered
                    logging.error(f"Referenced table {referenced_table} not yet populated for {table}.{column_name} (NOT NULL)")

                    # Generate a placeholder value
                    placeholder = self._generate_placeholder_value(col)
                    row_data[column_name] = placeholder
                    logging.info(f"Generated placeholder value {placeholder} for {table}.{column_name} -> {referenced_table}.{referenced_column}")
            else:
                # Not a foreign key, generate a value based on column type
                row_data[column_name] = self.data_gen.generate_value(col, table_name=table)

        return row_data

    def _generate_placeholder_value(self, column_info):
        """
        Generate a placeholder value for a column when handling circular dependencies

        Args:
            column_info (dict): Column information

        Returns:
            object: Generated placeholder value
        """
        # For primary keys, try to generate a unique value
        if column_info['column_key'] == 'PRI':
            data_type = column_info['data_type'].lower()

            if data_type in ('int', 'bigint', 'smallint', 'tinyint', 'mediumint'):
                # For integer primary keys, generate a large random number
                return random.randint(1000000, 9999999)
            elif data_type in ('varchar', 'char'):
                # For string primary keys, generate a UUID-like string
                return f"temp-{random.randint(100000, 999999)}"

        # For other columns, generate a normal value
        return self.data_gen.generate_value(column_info, table_name=None)

    def _handle_failed_tables(self):
        """
        Try to populate failed tables by trying random orders
        """
        logging.info("Attempting to resolve failed tables with random ordering")

        # Copy the failed tables set
        remaining_tables = self.failed_tables.copy()
        retry_count = 0

        while remaining_tables and retry_count < self.max_retries:
            retry_count += 1
            logging.info(f"Retry {retry_count}/{self.max_retries} for failed tables")

            # Try tables in random order
            tables_to_try = list(remaining_tables)
            random.shuffle(tables_to_try)

            # Track tables that were successfully populated in this iteration
            success_in_iteration = set()

            for table in tables_to_try:
                if self._populate_table(table, handle_circular=True):
                    success_in_iteration.add(table)

            # Remove successfully populated tables from the remaining set
            remaining_tables -= success_in_iteration

            # If no progress was made in this iteration, try a different approach
            if not success_in_iteration and remaining_tables:
                logging.info("No progress made in this iteration, trying partial population")

                # Try to insert at least one record in each remaining table
                for table in list(remaining_tables):
                    if self._try_partial_population(table):
                        remaining_tables.remove(table)

        # Update the failed tables set
        self.failed_tables = remaining_tables

    def _try_partial_population(self, table):
        """
        Try to populate a table with at least one record

        Args:
            table (str): Table name

        Returns:
            bool: Success status
        """
        logging.info(f"Attempting partial population of table: {table}")

        # Get column information
        columns = self.schema.table_columns[table]
        column_names = [col['column_name'] for col in columns]

        # Special handling for SyncSource table which has a column named 'Lag' (MySQL reserved keyword)
        if table == 'SyncSource':
            logging.info("Special handling for SyncSource table with reserved keyword columns")
            # Escape column names with backticks to handle reserved keywords
            escaped_column_names = [f"`{col_name}`" for col_name in column_names]

            # Prepare placeholders for SQL query
            placeholders = ', '.join(['%s'] * len(column_names))

            # Build INSERT query with escaped column names
            insert_query = f"INSERT INTO `{table}` ({', '.join(escaped_column_names)}) VALUES ({placeholders})"

            logging.debug(f"Using escaped column names for SyncSource: {insert_query}")
        else:
            # Prepare placeholders for SQL query
            placeholders = ', '.join(['%s'] * len(column_names))

            # Build INSERT query
            insert_query = f"INSERT INTO {table} ({', '.join(column_names)}) VALUES ({placeholders})"

        # Try multiple times with different random values
        for attempt in range(10):  # Try up to 10 times
            try:
                # Generate row data with special handling for circular dependencies
                row_data = self._generate_row_data(table, columns, handle_circular=True)

                # Execute insert
                if self.db.execute_query(insert_query, tuple(row_data.values()), commit=True):
                    # Get the inserted ID if there's an auto-increment column
                    for col in columns:
                        if col['extra'] == 'auto_increment':
                            last_id_query = "SELECT LAST_INSERT_ID() as id"
                            result = self.db.execute_query(last_id_query)
                            if result and result[0]['id']:
                                row_data[col['column_name']] = result[0]['id']

                    # Store inserted data for foreign key references
                    self.inserted_data[table].append(row_data)
                    logging.info(f"Successfully inserted one record into {table}")
                    return True

            except Error as e:
                logging.debug(f"Attempt {attempt+1} failed for {table}: {e}")
                # Continue with next attempt

        logging.error(f"Failed to insert even one record into {table} after multiple attempts")
        return False

File: internal/populator/populator.go
package populator

import (
	"fmt"
	"strings"
	"time"

	"github.com/sirupsen/logrus"
	"github.com/vitebski/mysql-dummy-populator/internal/analyzer"
	"github.com/vitebski/mysql-dummy-populator/internal/connector"
	"github.com/vitebski/mysql-dummy-populator/internal/generator"
	"github.com/vitebski/mysql-dummy-populator/pkg/models"
)

// DatabasePopulator populates database tables with fake data
type DatabasePopulator struct {
	DB             *connector.DatabaseConnector
	SchemaAnalyzer *analyzer.SchemaAnalyzer
	DataGenerator  *generator.DataGenerator
	NumRecords     int
	MaxRetries     int
	InsertedData   map[string][]map[string]interface{}
	FailedTables   map[string]bool
	Logger         *logrus.Logger
}

// NewDatabasePopulator creates a new database populator
func NewDatabasePopulator(
	db *connector.DatabaseConnector,
	schemaAnalyzer *analyzer.SchemaAnalyzer,
	dataGenerator *generator.DataGenerator,
	numRecords int,
	maxRetries int,
	logger *logrus.Logger,
) *DatabasePopulator {
	return &DatabasePopulator{
		DB:             db,
		SchemaAnalyzer: schemaAnalyzer,
		DataGenerator:  dataGenerator,
		NumRecords:     numRecords,
		MaxRetries:     maxRetries,
		InsertedData:   make(map[string][]map[string]interface{}),
		FailedTables:   make(map[string]bool),
		Logger:         logger,
	}
}

// PopulateDatabase populates the database with fake data
func (dp *DatabasePopulator) PopulateDatabase() bool {
	// Get table insertion order
	orderedTables, circularTables := dp.SchemaAnalyzer.GetTableInsertionOrder()

	// Track overall success
	success := true

	// Populate tables in order
	for _, table := range orderedTables {
		tableSuccess := false

		// Check if this table is part of a circular dependency
		isCircular := circularTables[table]

		if isCircular {
			// Handle circular dependency with special approach
			tableSuccess = dp.populateCircularTable(table)
		} else {
			// Normal table population
			tableSuccess = dp.populateTable(table)
		}

		if !tableSuccess {
			dp.FailedTables[table] = true
			success = false
		}
	}

	return success
}

// populateTable populates a single table with fake data
func (dp *DatabasePopulator) populateTable(table string) bool {
	dp.Logger.Infof("Populating table: %s", table)

	// Get columns for this table
	columns := dp.SchemaAnalyzer.TableColumns[table]
	if len(columns) == 0 {
		dp.Logger.Errorf("No columns found for table: %s", table)
		return false
	}

	// Check if this is a many-to-many table
	isManyToMany := dp.SchemaAnalyzer.ManyToManyTables[table]

	// Get foreign keys for this table
	foreignKeys := dp.SchemaAnalyzer.ForeignKeys[table]

	// Prepare column names and placeholders for the INSERT statement
	var columnNames []string
	var placeholders []string
	var columnObjects []models.Column

	for _, column := range columns {
		// Skip auto-increment columns
		if strings.Contains(strings.ToLower(column.Extra), "auto_increment") {
			continue
		}

		columnNames = append(columnNames, column.Name)
		placeholders = append(placeholders, "?")
		columnObjects = append(columnObjects, column)
	}

	if len(columnNames) == 0 {
		dp.Logger.Warningf("No insertable columns found for table: %s", table)
		return true // Consider this a success since there's nothing to insert
	}

	// Prepare the INSERT statement
	insertSQL := fmt.Sprintf(
		"INSERT INTO %s (%s) VALUES (%s)",
		table,
		strings.Join(columnNames, ", "),
		strings.Join(placeholders, ", "),
	)

	// Determine how many records to insert
	numRecords := dp.NumRecords
	if isManyToMany {
		// For many-to-many tables, calculate based on related tables
		numRecords = dp.calculateManyToManyRecords(table, foreignKeys)
	}

	// Generate and insert data
	var paramsList [][]interface{}
	var insertedRecords []map[string]interface{}

	for i := 0; i < numRecords; i++ {
		// Generate a record
		record, params := dp.generateRecord(table, columnNames, columnObjects, foreignKeys)
		
		if params != nil {
			paramsList = append(paramsList, params)
			insertedRecords = append(insertedRecords, record)
		}

		// Insert in batches of 100 records
		if len(paramsList) >= 100 || (i == numRecords-1 && len(paramsList) > 0) {
			_, err := dp.DB.ExecuteMany(insertSQL, paramsList)
			if err != nil {
				dp.Logger.Errorf("Error inserting data into table %s: %v", table, err)
				return false
			}

			// Store inserted data for reference
			dp.InsertedData[table] = append(dp.InsertedData[table], insertedRecords...)

			// Reset for next batch
			paramsList = nil
			insertedRecords = nil
		}
	}

	dp.Logger.Infof("Successfully populated table %s with %d records", table, numRecords)
	return true
}

// populateCircularTable populates a table involved in circular dependencies
func (dp *DatabasePopulator) populateCircularTable(table string) bool {
	dp.Logger.Infof("Populating circular dependency table: %s", table)

	// Get columns for this table
	columns := dp.SchemaAnalyzer.TableColumns[table]
	if len(columns) == 0 {
		dp.Logger.Errorf("No columns found for table: %s", table)
		return false
	}

	// Get foreign keys for this table
	foreignKeys := dp.SchemaAnalyzer.ForeignKeys[table]

	// Identify circular foreign keys
	var circularFKs []models.ForeignKey
	var nonCircularFKs []models.ForeignKey
	circularTables, _ := dp.SchemaAnalyzer.GetTableInsertionOrder()
	circularTablesMap := make(map[string]bool)
	for _, t := range circularTables {
		circularTablesMap[t] = true
	}

	for _, fk := range foreignKeys {
		if circularTablesMap[fk.ReferencedTable] {
			circularFKs = append(circularFKs, fk)
		} else {
			nonCircularFKs = append(nonCircularFKs, fk)
		}
	}

	// Prepare column names and placeholders for the INSERT statement
	var columnNames []string
	var placeholders []string
	var columnObjects []models.Column

	for _, column := range columns {
		// Skip auto-increment columns
		if strings.Contains(strings.ToLower(column.Extra), "auto_increment") {
			continue
		}

		columnNames = append(columnNames, column.Name)
		placeholders = append(placeholders, "?")
		columnObjects = append(columnObjects, column)
	}

	if len(columnNames) == 0 {
		dp.Logger.Warningf("No insertable columns found for table: %s", table)
		return true // Consider this a success since there's nothing to insert
	}

	// Prepare the INSERT statement
	insertSQL := fmt.Sprintf(
		"INSERT INTO %s (%s) VALUES (%s)",
		table,
		strings.Join(columnNames, ", "),
		strings.Join(placeholders, ", "),
	)

	// First pass: Insert records with NULL for circular foreign keys
	dp.Logger.Infof("First pass: Inserting records with NULL for circular foreign keys")
	var paramsList [][]interface{}
	var insertedRecords []map[string]interface{}

	for i := 0; i < dp.NumRecords; i++ {
		// Generate a record with NULL for circular foreign keys
		record, params := dp.generateRecordWithNullCircularFKs(table, columnNames, columnObjects, nonCircularFKs, circularFKs)
		
		if params != nil {
			paramsList = append(paramsList, params)
			insertedRecords = append(insertedRecords, record)
		}

		// Insert in batches of 100 records
		if len(paramsList) >= 100 || (i == dp.NumRecords-1 && len(paramsList) > 0) {
			_, err := dp.DB.ExecuteMany(insertSQL, paramsList)
			if err != nil {
				dp.Logger.Errorf("Error inserting data into table %s (first pass): %v", table, err)
				return false
			}

			// Store inserted data for reference
			dp.InsertedData[table] = append(dp.InsertedData[table], insertedRecords...)

			// Reset for next batch
			paramsList = nil
			insertedRecords = nil
		}
	}

	// Second pass: Update records with valid foreign keys
	dp.Logger.Infof("Second pass: Updating records with valid circular foreign keys")
	for _, fk := range circularFKs {
		// Skip if the referenced table has no data
		if len(dp.InsertedData[fk.ReferencedTable]) == 0 {
			dp.Logger.Warningf("Referenced table %s has no data, skipping update for %s.%s",
				fk.ReferencedTable, table, fk.Column)
			continue
		}

		// Get primary key column for this table
		var pkColumn string
		for _, col := range columns {
			if col.ColumnKey == "PRI" {
				pkColumn = col.Name
				break
			}
		}

		if pkColumn == "" {
			dp.Logger.Warningf("No primary key found for table %s, skipping update", table)
			continue
		}

		// Update each record with a random value from the referenced table
		for _, record := range dp.InsertedData[table] {
			// Get a random record from the referenced table
			referencedRecords := dp.InsertedData[fk.ReferencedTable]
			if len(referencedRecords) == 0 {
				continue
			}

			// Get the primary key value for this record
			pkValue := record[pkColumn]
			if pkValue == nil {
				continue
			}

			// Get a random referenced value
			referencedRecord := referencedRecords[time.Now().Nanosecond()%len(referencedRecords)]
			referencedValue := referencedRecord[fk.ReferencedColumn]
			if referencedValue == nil {
				continue
			}

			// Update the record
			updateSQL := fmt.Sprintf(
				"UPDATE %s SET %s = ? WHERE %s = ?",
				table,
				fk.Column,
				pkColumn,
			)

			_, err := dp.DB.ExecuteStatement(updateSQL, referencedValue, pkValue)
			if err != nil {
				dp.Logger.Errorf("Error updating circular foreign key %s.%s: %v", table, fk.Column, err)
				// Continue with other records
			}
		}
	}

	dp.Logger.Infof("Successfully populated circular dependency table %s with %d records", table, dp.NumRecords)
	return true
}

// generateRecord generates a single record for a table
func (dp *DatabasePopulator) generateRecord(
	table string,
	columnNames []string,
	columns []models.Column,
	foreignKeys []models.ForeignKey,
) (map[string]interface{}, []interface{}) {
	record := make(map[string]interface{})
	var params []interface{}

	// Create a map of foreign key columns for quick lookup
	fkMap := make(map[string]models.ForeignKey)
	for _, fk := range foreignKeys {
		fkMap[fk.Column] = fk
	}

	// Generate data for each column
	for i, columnName := range columnNames {
		column := columns[i]
		var value interface{}

		// Check if this is a foreign key
		if fk, isFk := fkMap[columnName]; isFk {
			// Get a random value from the referenced table
			value = dp.getRandomForeignKeyValue(fk)
			
			// If no value is available and the column is NOT NULL, this is a problem
			if value == nil && !column.IsNullable {
				dp.Logger.Errorf("No value available for NOT NULL foreign key %s.%s referencing %s.%s",
					table, columnName, fk.ReferencedTable, fk.ReferencedColumn)
				return nil, nil
			}
		} else {
			// Generate a value based on column type
			value = dp.DataGenerator.GenerateData(table, column)
		}

		record[columnName] = value
		params = append(params, value)
	}

	return record, params
}

// generateRecordWithNullCircularFKs generates a record with NULL values for circular foreign keys
func (dp *DatabasePopulator) generateRecordWithNullCircularFKs(
	table string,
	columnNames []string,
	columns []models.Column,
	nonCircularFKs []models.ForeignKey,
	circularFKs []models.ForeignKey,
) (map[string]interface{}, []interface{}) {
	record := make(map[string]interface{})
	var params []interface{}

	// Create maps for foreign key columns
	nonCircularFKMap := make(map[string]models.ForeignKey)
	for _, fk := range nonCircularFKs {
		nonCircularFKMap[fk.Column] = fk
	}

	circularFKMap := make(map[string]models.ForeignKey)
	for _, fk := range circularFKs {
		circularFKMap[fk.Column] = fk
	}

	// Generate data for each column
	for i, columnName := range columnNames {
		column := columns[i]
		var value interface{}

		// Check if this is a non-circular foreign key
		if fk, isFk := nonCircularFKMap[columnName]; isFk {
			// Get a random value from the referenced table
			value = dp.getRandomForeignKeyValue(fk)
			
			// If no value is available and the column is NOT NULL, this is a problem
			if value == nil && !column.IsNullable {
				dp.Logger.Errorf("No value available for NOT NULL foreign key %s.%s referencing %s.%s",
					table, columnName, fk.ReferencedTable, fk.ReferencedColumn)
				return nil, nil
			}
		} else if _, isCircularFK := circularFKMap[columnName]; isCircularFK {
			// Set circular foreign keys to NULL for now
			if !column.IsNullable {
				// If the column is NOT NULL, we need to handle it differently
				// For the first pass, we'll use a temporary value that will be updated later
				// This might violate constraints temporarily but will be fixed in the second pass
				value = dp.DataGenerator.GenerateData(table, column)
			} else {
				value = nil
			}
		} else {
			// Generate a value based on column type
			value = dp.DataGenerator.GenerateData(table, column)
		}

		record[columnName] = value
		params = append(params, value)
	}

	return record, params
}

// getRandomForeignKeyValue gets a random value from a referenced table
func (dp *DatabasePopulator) getRandomForeignKeyValue(fk models.ForeignKey) interface{} {
	// Check if we have inserted data for the referenced table
	referencedRecords, ok := dp.InsertedData[fk.ReferencedTable]
	if !ok || len(referencedRecords) == 0 {
		return nil
	}

	// Get a random record
	randomIndex := time.Now().Nanosecond() % len(referencedRecords)
	randomRecord := referencedRecords[randomIndex]

	// Return the referenced column value
	return randomRecord[fk.ReferencedColumn]
}

// calculateManyToManyRecords calculates how many records to insert for a many-to-many table
func (dp *DatabasePopulator) calculateManyToManyRecords(table string, foreignKeys []models.ForeignKey) int {
	// Get unique referenced tables
	referencedTables := make(map[string]bool)
	for _, fk := range foreignKeys {
		referencedTables[fk.ReferencedTable] = true
	}

	// Calculate based on the number of records in referenced tables
	var totalPossibleCombinations int = 1
	var availableReferencedTables int = 0

	for refTable := range referencedTables {
		refRecords, ok := dp.InsertedData[refTable]
		if ok && len(refRecords) > 0 {
			totalPossibleCombinations *= len(refRecords)
			availableReferencedTables++
		}
	}

	// If not all referenced tables have data, return 0
	if availableReferencedTables < len(referencedTables) {
		return 0
	}

	// Calculate a reasonable number of records
	// Use the smaller of: total possible combinations or 2*NumRecords
	if totalPossibleCombinations > 2*dp.NumRecords {
		return 2 * dp.NumRecords
	}
	return totalPossibleCombinations
}
